### PR TITLE
[rllib] Added evaluation script to RLLib

### PR DESCRIPTION
**Pull request by pschafhalter**
## What do these changes do?

- Added evaluation script to RLLib
- Added test for evaluation script

## Related issue number

- #1280

### CODE REVIEW

File: test/jenkins_tests/run_multi_node_tests.sh [hunk 1]
```diff
@@ -139,6 +139,9 @@ docker run --rm --shm-size=10G --memory=10G $DOCKER_SHA \
     --stop '{"training_iteration": 2}' \
     --config '{"num_workers": 2, "use_lstm": false, "use_pytorch": true, "model": {"grayscale": true, "zero_mean": false, "dim": 80, "channel_major": true}}'
 
+docker run --rm --shm-size=10G --memory=10G $DOCKER_SHA \
+  sh /ray/test/jenkins_tests/multi_node_tests/test_rllib_eval.sh
```
Comment (ericl) [inline test/jenkins_tests/run_multi_node_tests.sh]: nice!

File: python/ray/rllib/eval.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+required_named.add_argument(
+        "--run", type=str, required=True,
+        help="The algorithm or model to train. This may refer to the name "
+        "of a built-on algorithm (e.g. RLLib's DQN or PPO), or a "
+        "user-defined trainable function or class registered in the "
+        "tune registry.")
+required_named.add_argument(
+    "--env", type=str, required=True, help="The gym environment to use.")
+parser.add_argument(
+    "--hide", default=False, action="store_const", const=True,
```
Comment (ericl) [inline python/ray/rllib/eval.py]: --no-render

File: python/ray/rllib/eval.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+import argparse
+import gym
+import ray
+
+from agent import get_agent_class
+
+
+EXAMPLE_USAGE = """
+example usage:
+    ./train.py /tmp/ray/checkpoint_dir/checkpoint-0 --run DQN --env CartPole-v0
```
Comment (ericl) [inline python/ray/rllib/eval.py]: This should say ./eval.py right?

File: python/ray/rllib/eval.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env python
+
+import argparse
+import gym
+import ray
+
+from agent import get_agent_class
```
Comment (ericl) [inline python/ray/rllib/eval.py]: from ray.tune.agent import

File: python/ray/rllib/eval.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env python
+
```
Comment (ericl) [inline python/ray/rllib/eval.py]: Could you do the import __future__ stuff from train.py as well here?

File: python/ray/rllib/eval.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+"""
+
+
+parser = argparse.ArgumentParser(
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+        description="Evaluates a reinforcement learning agent "
+        "given a checkpoint.", epilog=EXAMPLE_USAGE)
+
+parser.add_argument("checkpoint", type=str,
+                    help="Checkpoint from which to evaluate.")
```
Comment (ericl) [inline python/ray/rllib/eval.py]: should make the indentation consistent

File: python/ray/rllib/eval.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+required_named = parser.add_argument_group("required named arguments")
+required_named.add_argument(
+        "--run", type=str, required=True,
+        help="The algorithm or model to train. This may refer to the name "
+        "of a built-on algorithm (e.g. RLLib's DQN or PPO), or a "
+        "user-defined trainable function or class registered in the "
+        "tune registry.")
+required_named.add_argument(
+    "--env", type=str, required=True, help="The gym environment to use.")
```
Comment (ericl) [inline python/ray/rllib/eval.py]: Do we also need --config?

File: python/ray/rllib/eval.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    ray.init()
+
+    cls = get_agent_class(args.run)
+    agent = cls(env=args.env)
+    agent._restore(args.checkpoint)
+
+    env = gym.make(args.env)
+    state = env.reset()
+    done = False
+    while not done:
```
Comment (ericl) [inline python/ray/rllib/eval.py]: Consider doing it in an infinite loop.

File: python/ray/rllib/eval.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    help="Surpress rendering of the environment.")
+
+
+if __name__ == "__main__":
+    args = parser.parse_args()
+    ray.init()
+
+    cls = get_agent_class(args.run)
+    agent = cls(env=args.env)
+    agent._restore(args.checkpoint)
```
Comment (richardliaw) [inline python/ray/rllib/eval.py]: `agent.restore` is the proper way

**Comment by pcmoritz** [comment]
@pschafhalter Can you rebase this on the latest master so we can see if it passes the Jenkins test you added?

**Comment by pschafhalter** [comment]
@pcmoritz it should be rebased now

File: python/ray/rllib/eval.py (hunk 5 above)
Comment (pschafhalter) [reply python/ray/rllib/eval.py]: Should be fixed in the latest commit

**Review by pcmoritz — COMMENTED** [review]
@pschafhalter LGTM, now that https://github.com/ray-project/ray/issues/1294 is fixed, can you re-introduce the loop in the shell script? All checkpoints should now start with index 1.

**Comment by pschafhalter** [comment]
@pcmoritz are checkpoints for AC3 fixed as well?

**Comment by pcmoritz** [comment]
@pschafhalter No, they are not I think. Do you want to create a github issue about it?

**Comment by pschafhalter** [comment]
@pcmoritz sure, I'll report it

**Review by ericl — COMMENTED** [review]
LGTM. Do you also want to update the docs (rllib.rst)?

**Comment by pschafhalter** [comment]
@ericl yes, I'll update the docs as well. Should I create a new section or document the script under Getting Started?

**Comment by ericl** [comment]
Let's integrate it into getting started

**Comment by pschafhalter** [comment]
@ericl I added some documentation. Let me know if there's anything I should change.

File: doc/source/rllib.rst [hunk 10]
```diff
@@ -68,6 +68,35 @@ The most important options are for choosing the environment
 with ``--env`` (any OpenAI gym environment including ones registered by the user
 can be used) and for choosing the algorithm with ``--run``
 (available options are ``PPO``, ``A3C``, ``ES`` and ``DQN``).
+In order to save checkpoints from which to restore training
```
Comment (ericl) [inline doc/source/rllib.rst]: Make this a subsection ("Evaluating trained agents") and move to below the specifying params section below?

**Review by ericl — APPROVED** [review]
Lgtm, one suggestion on organization since the added text is quite long.

**Pull request merged**